### DISCUSSION TITLE
Access agent state inside an agent function

### DISCUSSION THREAD

**Discussion by daniele-baccega** [FLAMEGPU/FLAMEGPU2#1107] [Q&A] [answered]
Hi,
is it possible to access to the agent state inside an agent function?

Thank you

**Comment by Robadob** [comment, marked as answer]
Not currently.

It should be relatively easy for us to add though, we'd just need to decide a suitable API. Agent state being a string, we'd want to avoid returning a `char*` that needs to be compared in device code as this would be unintuitive for users (e.g. they might start comparing pointers).

I'll create a new issue, and may be able to address it next week after it's been discussed with the team on Friday.

**Reply by ZHUOYUYUIO** [reply]
Hi!

I'm currently working on my ABM project using PyFLAMEGPU and ran into the same issue that's been discussed before.

I wanted to ask: Right now we only have the method b = pyflamegpu.isState("bar") for state checking. Is there no function like state = pyflamegpu.getState() to directly retrieve the current state variable?

Thanks for your help!

**Reply by Robadob** [reply]
Sorry there's been no progress on this issue. 


Based on how `isState()` is implemented.

https://github.com/FLAMEGPU/FLAMEGPU2/blob/8b68331043243be3fb28e07bd531ea9d95c947a3/include/flamegpu/runtime/detail/curve/DeviceCurve.cuh#L527

Technically in the C++ API it may be possible to retrieve it with

```c++
using detail::sm;
char* state = sm()->state_name;
```

But as previously stated, this is a `char*` (a character array) rather than a standard string.

What's your requirement for having it in string form?

**Reply by ZHUOYUYUIO** [reply]
In my 3D space-based project, if the number of nearby agents (within a certain space range) in state A exceeds a given threshold, the current agent will also switch to state A.

**Reply by ZHUOYUYUIO** [reply]
Perhaps I could implement this by directly setting the variable or using the isState method. 
Additionally, I’d like to confirm: in b = pyflamegpu.isState("bar"), does b return a boolean (True/False)?

**Comment by Robadob** [comment]
It's probably easiest for you to store your own state variable in agents in
that case.

Yes isState() returns true/false

**Comment by Robadob** [comment]
You couldn't implement different message output functions per state, which
set a flag differently. This is practical assuming you only have a few
states.

**Reply by ZHUOYUYUIO** [reply]
Thanks for the help—I’ll test both methods (direct variable vs. isState) and compare them.
